// @nrwl/linter/src/executors/eslint/hasher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const app_root_1 = require("@nrwl/tao/src/utils/app-root");
const workspace_1 = require("@nrwl/tao/src/shared/workspace");
function run(task, taskGraph, hasher) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function* () {
        if (task.overrides['hasTypeAwareRules'] === true) {
            return hasher.hashTaskWithDepsAndContext(task);
        }
        const command = hasher.hashCommand(task);
        const sources = yield hasher.hashSource(task);
        const deps = allDeps(task.id, taskGraph);
        const workspace = new workspace_1.Workspaces(app_root_1.appRootPath).readWorkspaceConfiguration();
        const tags = hasher.hashArray(deps.map((d) => (workspace.projects[d].tags || []).join('|')));
        const context = yield hasher.hashContext();
        return {
            value: hasher.hashArray([
                command,
                sources,
                tags,
                context.implicitDeps.value,
                context.runtime.value,
            ]),
            details: {
                command,
                nodes: { [task.target.project]: sources, tags },
                implicitDeps: context.implicitDeps.files,
                runtime: context.runtime.runtime,
            },
        };
    });
}
exports.default = run;
function allDeps(taskId, taskGraph) {
    return [
        ...taskGraph.dependencies[taskId].map((task) => taskGraph.tasks[task].target.project),
        ...taskGraph.dependencies[taskId].flatMap((task) => allDeps(task, taskGraph)),
    ];
}
